src/test/app/Offer_test.cpp

```diff
@@ -795,22 +850,831 @@ class Offer_test : public beast::unit_test::suite
+        auto const alice = Account {"alice"};
+        auto const bob = Account {"bob"};
+        auto const USD = gw["USD"];
+
+        env.fund (XRP (100000), gw, alice, bob);
+
+        env (trust (alice, USD (1000)));
+
+        env (pay (gw, alice, alice["USD"] (500)));
+
+        env (offer (alice, XRP (150000), USD (50)));
+        env (offer (bob, USD (1), XRP (3000)));
+
+        auto jrr = ledgerEntryState (env, bob, gw, "USD");
+        BEAST_EXPECT(jrr[jss::node][sfBalance.fieldName][jss::value] == "-1");
+        jrr = ledgerEntryRoot(env, bob);
+        BEAST_EXPECT(
+            jrr[jss::node][sfBalance.fieldName] ==
+            std::to_string(
+                XRP (100000).value ().mantissa () -
+                XRP (3000).value ().mantissa () -
+                env.current ()->fees ().base * 1)
+        );
+
+        jrr = ledgerEntryState (env, alice, gw, "USD");
+        BEAST_EXPECT(jrr[jss::node][sfBalance.fieldName][jss::value] == "-499");
+        jrr = ledgerEntryRoot (env, alice);
+        BEAST_EXPECT(
+            jrr[jss::node][sfBalance.fieldName] ==
+            std::to_string(
+                XRP (100000).value ().mantissa () +
+                XRP (3000).value ().mantissa () -
+                env.current ()->fees ().base * 2)
+        );
+    }
+
+    void
+    testcaseOfferAcceptThenCancel()
+    {
+        testcase ("Offer Accept then Cancel.");
+
+        using namespace jtx;
+
+        Env env {*this};
+        auto const USD = env.master["USD"];
+
+        auto const nextOfferSeq = env.seq (env.master);
+        env (offer (env.master, XRP (500), USD (100)));
+        env.close ();
+
+        Json::Value cancelOffer;
+        cancelOffer[jss::Account] = env.master.human();
+        cancelOffer[jss::OfferSequence] = nextOfferSeq;
+        cancelOffer[jss::TransactionType] = "OfferCancel";
+        env (cancelOffer);
+        BEAST_EXPECT(env.seq (env.master) == nextOfferSeq + 2);
+
+        // ledger_accept, call twice and verify no odd behavior
+        env.close();
+        env.close();
+        BEAST_EXPECT(env.seq (env.master) == nextOfferSeq + 2);
+    }
+
+    void
+    testcaseOfferCancelPastAndFuture()
+    {
+
+        testcase ("Offer Cancel Past and Future Sequence.");
+
+        using namespace jtx;
+
+        Env env {*this};
+        auto const alice = Account {"alice"};
+
+        auto const nextOfferSeq = env.seq (env.master);
+        env.fund (XRP (10000), alice);
+
+        Json::Value cancelOffer;
+        cancelOffer[jss::Account] = env.master.human();
+        cancelOffer[jss::OfferSequence] = nextOfferSeq;
+        cancelOffer[jss::TransactionType] = "OfferCancel";
+        env (cancelOffer);
+
+        cancelOffer[jss::OfferSequence] = env.seq (env.master);
+        env (cancelOffer, ter(temBAD_SEQUENCE));
+
+        cancelOffer[jss::OfferSequence] = env.seq (env.master) + 1;
+        env (cancelOffer, ter(temBAD_SEQUENCE));
+
+        env.close();
```

Comment:
There's probably no harm, but I don't think the calls to `env.close` do any good here since you don't look at the ledger after calling `close`.